ddtrace/tracer/abandoned_spans.go

```diff
@@ -0,0 +1,224 @@
+	if !ok || s == nil {
+		return nil, false
+	}
+	return s, true
+}
+
+// reportAbandonedSpans periodically finds and reports potentially abandoned
+// spans that are older than the given interval. These spans are stored in a
+// bucketed linked list, sorted by their `Start` time, where the front of the
+// list contains the oldest spans, and the end of the list contains the newest spans.
+func (t *tracer) reportAbandonedSpans(interval time.Duration) {
+	tick := time.NewTicker(tickerInterval)
+	defer tick.Stop()
+	for {
+		select {
+		case <-tick.C:
+			logAbandonedSpans(t.abandonedSpans, &interval)
+		case s := <-t.cIn:
+			bNode := t.abandonedSpans.Front()
+			if bNode == nil {
+				b := list.New()
+				b.PushBack(s)
+				t.abandonedSpans.PushBack(b)
+				break
+			}
+			// All spans within the same bucket should have a start time that
+			// is within `interval` nanoseconds of each other.
+			// This loop should continue until it finds the an existing bucket
+			// with spans that have started within `interval` nanoseconds before
+			// the new span has started.
+			for bNode != nil {
+				bucket, bOk := isBucketNode(bNode)
+				if !bOk {
+					bNode = bNode.Next()
+					continue
+				}
+				sNode := bucket.Front()
+				if sNode == nil {
+					bNode = bNode.Next()
+					continue
+				}
+				sp, sOk := isSpanNode(sNode)
+				if sOk && s.Start-sp.Start <= interval.Nanoseconds() {
+					bucket.PushBack(s)
+					break
+				}
+				bNode = bNode.Next()
+			}
+			if bNode != nil {
+				break
+			}
+			// If no matching bucket exists, create a new one and append the new
+			// span to the top of the bucket.
+			b := list.New()
+			b.PushBack(s)
+			t.abandonedSpans.PushBack(b)
+		case s := <-t.cOut:
+			// This loop should continue until it finds the bucket with spans
+			// starting within `interval` nanoseconds of the finished span,
+			// then remove that span from the bucket.
+			for node := t.abandonedSpans.Front(); node != nil; node = node.Next() {
+				bucket, ok := isBucketNode(node)
+				if !ok {
+					continue
+				}
+				spNode := bucket.Front()
+				sp, ok := isSpanNode(spNode)
+				if !ok {
+					continue
+				}
+				if s.Start-sp.Start > interval.Nanoseconds() || sp.Start-s.Start > interval.Nanoseconds() {
+					continue
+				}
+
+				for spNode != nil {
+					sp, ok := isSpanNode(spNode)
+					if !ok {
+						continue
+					}
+					if s.SpanID == sp.SpanID {
+						bucket.Remove(spNode)
+						if bucket.Front() == nil {
+							t.abandonedSpans.Remove(node)
+						}
+						break
+					}
+					spNode = spNode.Next()
+				}
+			}
+		case <-t.stop:
+			logAbandonedSpans(t.abandonedSpans, nil)
+			return
+		}
+	}
+}
+
+// abandonedSpanString takes a span and returns a human readable string representing
+// that span. If `interval` is not nil, it will check if the span is older than the
+// user configured timeout, and return an empty string if it is not.
+func abandonedSpanString(s *span, interval *time.Duration, curTime int64) string {
```

Comment:
Any reason to not just use the already existing `String` method that exists for spans? (span.go L614)

Reply:
The existing span method doesn't really have the information I wanted in the logs; i.e. I could have used `"%v"` to print spans, but it didn't have the `age` or `name` tag already. I was also wary of having more information that needed (i.e. the `parent_id` tag), since this debug mode can print a _lot_ of spans, so I'm trying to save bytes wherever I can.